HLRC API for _termvectors

mayya-sharipova opened the pull request:
relates to #27205

elasticmachine commented:
Pinging @elastic/es-core-infra

nik9000 commented on client/rest-high-level/src/main/java/org/elasticsearch/client/core/TermVectorsRequest.java:
```diff
@@ -0,0 +1,230 @@
+        if (docBuilder != null) {
+            BytesReference doc = BytesReference.bytes(docBuilder);
+            try (InputStream stream = doc.streamInput()) {
+                builder.rawField("doc", stream, docBuilder.contentType());
```
I really don't like the name of this method! It makes me think we're copying the field raw without checking the content type but we *are* checking it.

nik9000 commented on client/rest-high-level/src/main/java/org/elasticsearch/client/core/TermVectorsResponse.java:
```diff
@@ -0,0 +1,452 @@
+    private final long docVersion;
+    private final boolean found;
+    private final long tookInMillis;
+    private List<TermVector> termVectorList = null;
```
It feels a little weird to have just this part be mutable.

hub-cap replied:
is it possible for the `ConstructingObjectParser` to pass in a `NamedObjects` `constructorArg`? Right now on line 67 is where its being mutated.

nik9000 commented on client/rest-high-level/src/main/java/org/elasticsearch/client/core/TermVectorsResponse.java:
```diff
@@ -0,0 +1,452 @@
+            }
+
+        }
+        }
```
Funky indentation.

nik9000 commented on client/rest-high-level/src/main/java/org/elasticsearch/client/core/TermVectorsResponse.java:
```diff
@@ -0,0 +1,452 @@
+            public Term() {};
+
+            public static Term fromXContent(XContentParser parser, String term) {
+                Term t = TERMPARSER.apply(parser, null);
```
Maybe put `TERMPARSER` into this class?

nik9000 replied:
The same for the other parsers, I think.

nik9000 commented on client/rest-high-level/src/main/java/org/elasticsearch/client/core/TermVectorsResponse.java:
```diff
@@ -0,0 +1,452 @@
+            private float score = -1f;
+            private List<Token> tokens = null;
+
+            public Term() {};
```
Why is this one mutable when the others are immutable? Maybe make all immutable?

hub-cap commented on client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java:
```diff
@@ -1122,6 +1123,26 @@ static Request analyze(AnalyzeRequest request) throws IOException {
 
+    static Request termVectors(TermVectorsRequest tvrequest) throws IOException {
+        String endpoint = "_termvectors";
+        if (tvrequest.getId() != null) {
```
the EndpointBuilder() gladly accepts nulls in its `addPathPart(...)` method. You dont need to worry about passing a null value in, as it just wont put it in the path.

hub-cap commented on client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java:
```diff
@@ -1122,6 +1123,26 @@ static Request analyze(AnalyzeRequest request) throws IOException {
+        }
+        Request request = new Request(HttpGet.METHOD_NAME, endpoint);
+        Params params = new Params(request);
+        if (tvrequest.getRouting() != null) params.withRouting(tvrequest.getRouting());
```
the params methods all call `putParam` and that does a null check, and will not add it if its null. So no need to check here.

hub-cap commented on client/rest-high-level/src/main/java/org/elasticsearch/client/core/TermVectorsRequest.java:
```diff
@@ -0,0 +1,230 @@
+
+        if (filterSettings != null) {
+            builder.startObject("filter");
+            if (filterSettings.containsKey("max_num_terms")) builder.field("max_num_terms", filterSettings.get("max_num_terms"));
```
maybe a helper method here to reduce the copy/paste

hub-cap commented on client/rest-high-level/src/main/java/org/elasticsearch/client/core/TermVectorsResponse.java:
```diff
@@ -0,0 +1,452 @@
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
```
I dont think we need this

hub-cap commented on client/rest-high-level/src/test/java/org/elasticsearch/client/core/TermVectorsResponseTests.java:
```diff
@@ -0,0 +1,124 @@
+import java.util.function.Predicate;
+
+
+public class TermVectorsResponseTests extends AbstractXContentTestCase<TermVectorsResponse> {
```
when removing the ToXContent, you will have to manually create a parser to test. This could be nasty on such a large class hierarchy. Im sorry, but its the best we have now. We need to come up with a better way. You should be able to move the code out of the response and just use it in a test.

hub-cap commented on server/src/main/java/org/elasticsearch/index/reindex/DeleteByQueryRequest.java:
```diff
@@ -61,7 +61,7 @@ public DeleteByQueryRequest(String... indices) {
     }
 
-    DeleteByQueryRequest(SearchRequest search) {
+    public DeleteByQueryRequest(SearchRequest search) {
```
Is this intended? I dont see any uses of it here.

mayya-sharipova commented:
@hub-cap  @nik9000 I have tried to address your comments, can you please continue the review.

mayya-sharipova commented on client/rest-high-level/src/main/java/org/elasticsearch/client/core/TermVectorsRequest.java:
```diff
@@ -0,0 +1,216 @@
+            String[] filterSettingNames =
+                {"max_num_terms", "min_term_freq", "max_term_freq", "min_doc_freq", "max_doc_freq", "min_word_length", "max_word_length"};
+            for (String settingName : filterSettingNames) {
+                if (filterSettings.containsKey(settingName)) builder.field(settingName, filterSettings.get(settingName));
```
All these parameters are optional, and a request doesn't require any of them, that is why I think it is fine to have it as a map.

mayya-sharipova commented:
@nik9000  Thanks for such a detailed feedback, Nik. I have addressed all your comments, can you please continue the review.

The only comment I have not addressed is  to make `filterSettings` a class in the `TermVectorsRequest`.  This is because as all the parameters in  `filterSettings` are optional, and a request doesn't require any of them, I think it is fine to have `filterSettings`  as a map.

hub-cap commented on client/rest-high-level/src/main/java/org/elasticsearch/client/core/TermVectorsRequest.java:
```diff
@@ -0,0 +1,229 @@
+     * @param docId - id of the document
+     */
+    public TermVectorsRequest(String index, String type, String docId) {
+        this.index = index;
```
```
this(index, type);
this.id = docId;
```

mayya-sharipova commented:
@hub-cap  thank you for the review.

@nik9000  Do you still want to make the final round of review, or I can merge this PR?
